src/pymatgen/core/structure.py

```diff
@@ -1098,22 +1099,24 @@ def __init__(
         self._properties = properties or {}
 
     def __eq__(self, other: object) -> bool:
+        """Define equality by comparing all three attributes: lattice, sites, properties."""
         needed_attrs = ("lattice", "sites", "properties")
 
-        # Return NotImplemented as in https://docs.python.org/3/library/functools.html#functools.total_ordering
         if not all(hasattr(other, attr) for attr in needed_attrs):
             return NotImplemented
 
-        # TODO (DanielYang59): fix below type
-        other = cast(Structure, other)  # make mypy happy
```

Comment:
I believe checking attribute (duck-typing-like) would better reflect the goal of the implementation, i.e. we don't need `other` to be a `Structure` type, we just need it to have certain attributes (`lattice`, `sites` and `properties`)

Reply:
I would revert this for now as I'm not 100% sure what is the best way to check for `in` and `len` support